### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by ram4nd**
These might be considered as minor performance changes, but together they might make a dent.

- array_push(...) calls behaving as '$array []= ...'. ' []= ' works faster than invoking functions in PHP.
- ￼If you only want to determine if a particular needle occurs within haystack, use the faster and less memory intensive function strpos() instead.
- These cases can be refactored by using case-sensitive functions strstr(...)/strpos() for better performance because matched patterns does not contain any characters.
- substr() shorthand can be used instead of strlen().
- A 'substr()' call can behave as array access, but hiding some error-handling logic behind as well. Using array access is a preferred way, because you have to implicitly write (read as document) errors handling code.
- Cascade str_replace() calls when the function has been applied to a variable sequentially.
- Fix count function in loops.

### CODE REVIEW

File: src/Barcode.php [hunk 1]
```diff
@@ -141,8 +141,14 @@ public function getBarcode($code, $type, $pr = 0.0)
 
 	private function sanitizeCode($code)
 	{
-		$code = str_replace(chr(194) . chr(160), ' ', $code); // mPDF 5.3.95  (for utf-8 encoded)
-		$code = str_replace(chr(160), ' ', $code); // mPDF 5.3.95	(for win-1252)
+		$code = str_replace(
+			array(
```
Comment (finwe) [inline src/Barcode.php]: Short arrays, please.

File: src/Barcode/EanUpc.php [hunk 2]
```diff
@@ -93,7 +93,7 @@ private function init($code, $length)
 			$invalidUpce = false;
 			if (($tmp == '000') or ($tmp == '100') or ($tmp == '200')) {
 				// Manufacturer code ends in 000, 100, or 200
-				$upceCode = substr($code, 2, 2) . substr($code, 9, 3) . substr($code, 4, 1);
+				$upceCode = substr($code, 2, 2) . substr($code, 9, 3) . $code[4];
```
Comment (finwe) [inline src/Barcode/EanUpc.php]: Possible undefined key.

File: src/Barcode/EanUpc.php [hunk 3]
```diff
@@ -106,16 +106,16 @@ private function init($code, $length)
 						$invalidUpce = true;
 					}
 				} else {
-					$tmp = substr($code, 6, 1);
+					$tmp = $code[6];
```
Comment (finwe) [inline src/Barcode/EanUpc.php]: Possible undefined key.

File: src/Barcode/Postnet.php [hunk 4]
```diff
@@ -65,14 +65,13 @@ private function init($code, $gapWidth, $planet)
 		$bararray = ['code' => $code, 'maxw' => 0, 'maxh' => 5, 'bcode' => []];
 
 		$k = 0;
-		$code = str_replace('-', '', $code);
-		$code = str_replace(' ', '', $code);
+		$code = str_replace(array('-', ' '), '', $code);
```
Comment (finwe) [inline src/Barcode/Postnet.php]: Short arrays.

File: src/Barcode/Postnet.php [hunk 5]
```diff
@@ -65,14 +65,13 @@ private function init($code, $gapWidth, $planet)
...
 		$len = strlen($code);
 
 		// calculate checksum
 		$sum = 0;
 		for ($i = 0; $i < $len; ++$i) {
-			$sum += (int) $code[$i];
+			$sum += (int)$code[$i];
```
Comment (finwe) [inline src/Barcode/Postnet.php]: Please keep spaces after cast.

File: src/Form.php [hunk 6]
```diff
@@ -382,7 +382,7 @@ function print_ob_select($objattr, $w, $h, $texto, $rtlalign, $k, $blockdir)
 			} // mPDF 5.3.37
 			$data = ['VAL' => [], 'OPT' => [], 'SEL' => [],];
 			if (isset($objattr['items'])) {
-				for ($i = 0; $i < count($objattr['items']); $i++) {
+				for ($i = 0, $iMax = count($objattr['items']); $i < $iMax; $i++) {
```
Comment (finwe) [inline src/Form.php]: It is best to move the count call in front of the for loop entirely.

**Comment by finwe** [comment]
You have still a 
> PHP Parse error:  syntax error, unexpected 'for' (T_FOR) in /home/finwe/www/_github/mpdf/repo/src/Otl.php on line 314

Also please fix the lint.

**Comment by finwe** [comment]
More notices, with example61_new_mPDF_v6-0_features.php:

> PHP Notice:  Undefined offset: 474 in src/Otl.php on line 1017

When making a change this vast, I really recommend testing it out on the examples repository. The generate-examples.sh script there allows you to point mPDF root to a path so that you don't need to rely on local composer.

**Comment by fulldecent** [comment]
@ram4nd Thank you for working on this. It is good to see some hand-tuning and learn some of these little habits.

**Comment by fglueck** [comment]
Great job!

**Comment by finwe** [comment]
@ram4nd Will you find time to finish this so it can be incorporated?

**Comment by finwe** [comment]
Closing as stale. Feel free to reopen with fixes, rebased onto current development.

**Pull request closed without merging**